src/braket/circuits/moments.py

```diff
@@ -119,35 +133,106 @@ def time_slices(self) -> Dict[int, List[Instruction]]:
+
+    def _add(self, instruction: Instruction, noise_index: int = 0) -> None:
+
+        if isinstance(instruction.operator, Noise):
+            self.add_noise(instruction)
+
+        else:
+            qubit_range = instruction.target
+            time = max([self._max_time_for_qubit(qubit) for qubit in qubit_range]) + 1
+
+            # Mark all qubits in qubit_range with max_time
+            for qubit in qubit_range:
+                self._max_times[qubit] = max(time, self._max_time_for_qubit(qubit))
+
+            self._moments[MomentsKey(time, instruction.target, "gate", noise_index)] = instruction
+            self._qubits.update(instruction.target)
+            self._depth = max(self._depth, time + 1)
+
+    def add_noise(
+        self, instruction: Instruction, input_type: str = "noise", noise_index: int = 0
+    ) -> None:
+        """
+        Add instructions to self.
+
+        Args:
+            instructions (Iterable[Instruction]): Instructions to add to self. The instruction is
+                added to the max time slice in which the instruction fits.
+        """
+
+        self._add_noise(instruction, input_type, noise_index)
+
+    def _add_noise(self, instruction: Instruction, input_type: str, noise_index: int) -> None:
 
-    def _add(self, instruction: Instruction) -> None:
         qubit_range = instruction.target
-        time = max([self._max_time_for_qubit(qubit) for qubit in qubit_range]) + 1
+        time = max(0, *[self._max_time_for_qubit(qubit) for qubit in qubit_range])
+        if input_type == "initialization_noise":
+            time = 0
+
+        while MomentsKey(time, qubit_range, input_type, noise_index) in self._moments:
+            noise_index = noise_index + 1
 
-        # Mark all qubits in qubit_range with max_time
-        for qubit in qubit_range:
-            self._max_times[qubit] = max(time, self._max_time_for_qubit(qubit))
+        self._moments[MomentsKey(time, qubit_range, input_type, noise_index)] = instruction
+        self._qubits.update(qubit_range)
 
-        self._moments[MomentsKey(time, instruction.target)] = instruction
-        self._qubits.update(instruction.target)
-        self._depth = max(self._depth, time + 1)
+    def sort_moments(self) -> None:
+        """
+        Make the disordered moments in order.
+
+        1. Make the readout noise in the end
+        2. Make the initialization noise at the beginning
+        """
+        # key for 'noise', 'gate' and 'gate_noise'
+        key_noise = []
+        # key for 'initialization_noise'
+        key_initialization_noise = []
+        # key for 'readout_noise'
+        key_readout_noise = []
+        moment_copy = OrderedDict()
+        sorted_moment = OrderedDict()
+
+        for key, instruction in self._moments.items():
+            moment_copy[key] = instruction
+            if key.moment_type == "readout_noise":
+                key_readout_noise.append(key)
+            elif key.moment_type == "initialization_noise":
```

Comment:
Rather than strings, create an enum:
```python
class MomentType(Enum, str):
    READOUT = "readout_noise"
    INITIALIZATION = "initialization_noise"
```
This way, we're no longer prone to misspellings or string changes.